# Distributed Lag Models - Koyck Transformation

- Salvatore, Dominic & Reagle, Derrick; Shaum's Outlines: Statistics and Econometrics; 2nd Edition; 2002; pg. 182-183

## Import Libraries

In [5]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
from statsmodels.sandbox.regression.predstd import wls_prediction_std

import matplotlib.pylab as plt
import matplotlib as mpl
%matplotlib inline
from matplotlib import rcParams
rcParams['figure.figsize'] = 15,6


- The premise for the Koyck transformation is that Y is sluggish to respond to a stimulus
- Types of stimulus might include: (policy change / advertising / medication)
- Therefore Y in period t is dependent on variable(s) observed in prior time periods. 
- The number of time periods may be known/small or they may need to be estimated.

## Import Data - create lag in inventories and drop the first observation

In [16]:
data = pd.read_csv('C://Users//Family//Downloads//KoyceData815.csv', index_col='Year')
data['constant'] = 1
data['Inventories_L1'] = data.Inventories.shift(1)
print(data.head())
data1 = data.iloc[1:19,:]
print(data1)
print('/n Data Types:')
print(data1.dtypes)

      Inventories  Sales  constant  Inventories_L1
Year                                              
1981          546    345         1             NaN
1982          574    344         1           546.0
1983          590    396         1           574.0
1984          650    417         1           590.0
1985          664    428         1           650.0
      Inventories  Sales  constant  Inventories_L1
Year                                              
1982          574    344         1           546.0
1983          590    396         1           574.0
1984          650    417         1           590.0
1985          664    428         1           650.0
1986          663    445         1           664.0
1987          710    473         1           663.0
1988          767    522         1           710.0
1989          815    533         1           767.0
1990          841    542         1           815.0
1991          835    542         1           841.0
1992          843    585       

## Estimate the model

In [17]:
model = sm.OLS(data1.Inventories, data1[['constant','Sales','Inventories_L1']])
results = model.fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:            Inventories   R-squared:                       0.994
Model:                            OLS   Adj. R-squared:                  0.993
Method:                 Least Squares   F-statistic:                     1152.
Date:                Tue, 25 Sep 2018   Prob (F-statistic):           3.81e-17
Time:                        20:32:32   Log-Likelihood:                -72.831
No. Observations:                  18   AIC:                             151.7
Df Residuals:                      15   BIC:                             154.3
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
constant          88.3716     24.248      3.

C:\Program Files (x86)\Microsoft Visual Studio\Shared\Anaconda3_64\lib\site-packages\scipy\stats\stats.py:1334: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=18
  "anyway, n=%i" % int(n))


- Lambda represents the geometric decline in weight from prior period.
- lambda = coeff(Inventories_L1) = .4996 = .50 ==> for period 1, the mean declines by 50% and continues to decline by same percentage for each subsequent time period.
- alpha ==> alpha*(1-lambda) = 88.3716  ==> alpha = 88.3716/(1-.50) = 176.7432

- The above model estimators are biased and inconsistent.
- The above model also assumes the fixed geometric decline (.50) which is rarely observed in the wild.

## Almon Lag Model